### PR TITLE
Updated muon mc classifier

### PR DESCRIPTION
**Pull request by abbiendi**
This is an update of the MC truth association tools for muons.
The association of reco::Muons has been updated to allow the association of (Global || Tracker) muons. The MuonMCClassifier has been updated and few bugs have been fixed.
The tools can be used to assess the performance of Muon ID and have been employed for the Particle-Flow performance paper (PRF-14-001).
The new code was initially developed in that context (in 7_4_X) and extensively tested there.
Now it has been ported to CMSSW_9_3_X and tested on MC events with "RelVal"-like content.
An example is provided, running on Phase2 events having the TrackingParticles in the event content:
 MuonAnalysis/MuonAssociators/test/MuonMCClassifier/testClassByHitsTP_cfg.py

This PR is relevant for performance studies (e.g. Phase2 Muon TDR) and in general for developments in the muon ID validation.

@calabria

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

File: MuonAnalysis/MuonAssociators/plugins/MuonMCClassifier.cc [hunk 1]
```diff
@@ -291,87 +301,166 @@ MuonMCClassifier::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
...
+	    tpBx[i]      = tp->eventId().bunchCrossing();
+	    tpEv[i]      = tp->eventId().event();
 
             hitsPdgId[i] = tp->pdgId();
             prodRho[i]   = tp->vertex().Rho();
             prodZ[i]     = tp->vertex().Z();
-	    edm::LogVerbatim("MuonMCClassifier") <<"\t TP pdgId = "<<hitsPdgId[i] << ", vertex rho = " << prodRho[i] << ", z = " << prodZ[i];
+
+	    // added info on GEANT process producing the TrackingParticle
+	    std::vector<SimVertex> G4Vs = tp->parentVertex()->g4Vertices();
```
Comment (davidlange6) [inline MuonAnalysis/MuonAssociators/plugins/MuonMCClassifier.cc]: hi @abbiendi - this can be const std::vector<SimVertex> &G4Vs to avoid an unneeded copy

Comment (abbiendi) [inline MuonAnalysis/MuonAssociators/plugins/MuonMCClassifier.cc]: sure!  thanks

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by smuzaffar** [comment]
code-checks

**Comment by abbiendi** [comment]
Ok, I am committing the suggested changes. In addition a small bug fix with minor effects.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by davidlange6** [comment]
merge

**Pull request merged**